packages/core/snapshot/manager.go

```diff
@@ -0,0 +1,175 @@
+	ecRecord, lastConfirmedEpoch, err := m.tangle.Options.CommitmentFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	// lock the entire ledger in notarization manager until the snapshot is created.
+	m.notarizationMgr.WriteLockLedger()
+	defer m.notarizationMgr.WriteUnlockLedger()
+
+	// lock the entire solid entry points storage until the snapshot is created.
+	m.sepsLock.Lock()
+	defer m.sepsLock.Unlock()
+
+	headerPord := func() (header *ledger.SnapshotHeader, err error) {
+		header = &ledger.SnapshotHeader{
+			FullEpochIndex: lastConfirmedEpoch,
+			DiffEpochIndex: ecRecord.EI(),
+			LatestECRecord: ecRecord,
+		}
+		return header, nil
+	}
+
+	sepsProd := NewSolidEntryPointsProducer(lastConfirmedEpoch, ecRecord.EI(), m)
+	outputWithMetadataProd := NewLedgerUTXOStatesProducer(lastConfirmedEpoch, m.notarizationMgr)
+	epochDiffsProd := NewEpochDiffsProducer(lastConfirmedEpoch, ecRecord.EI(), m.notarizationMgr)
+
+	header, err = CreateSnapshot(snapshotFileName, headerPord, sepsProd, outputWithMetadataProd, epochDiffsProd)
+
+	return
+}
+
+// LoadSolidEntryPoints add solid entry points to storage.
+func (m *Manager) LoadSolidEntryPoints(seps *SolidEntryPoints) {
+	if seps == nil {
+		return
+	}
+
+	for _, b := range seps.Seps {
+		m.insertSolidEntryPoint(b)
+	}
+}
+
+func (m *Manager) insertSolidEntryPoint(id tangleold.BlockID) error {
+	m.sepsLock.Lock()
+	defer m.sepsLock.Unlock()
+
+	sepsStore, err := m.baseStore.WithRealm(append([]byte{database.PrefixSnapshot, prefixSolidEntryPoint}, id.EpochIndex.Bytes()...))
+	if err != nil {
+		panic(err)
+	}
+
+	if err := sepsStore.Set(id.Bytes(), id.Bytes()); err != nil {
+		return errors.New("Fail to insert block to epoch store")
+	}
+
+	return nil
+}
+
+func (m *Manager) removeSolidEntryPoint(b *tangleold.Block, lastConfirmedEpoch epoch.Index) (err error) {
+	m.sepsLock.Lock()
+	defer m.sepsLock.Unlock()
+
+	sepsStore, err := m.baseStore.WithRealm(append([]byte{database.PrefixSnapshot, prefixSolidEntryPoint}, b.EI().Bytes()...))
+	if err != nil {
+		panic(err)
+	}
+
+	idBytes, err := sepsStore.Get(b.ID().Bytes())
+	if err != nil || idBytes == nil {
+		return errors.New("solid entry point doesn't exist in storage or fail to fetch it")
+	}
+
+	var blkID tangleold.BlockID
+	_, err = blkID.FromBytes(idBytes)
+	if err != nil {
+		return err
+	}
+
+	// cannot remove sep from confirmed epoch
+	if blkID.EpochIndex <= lastConfirmedEpoch {
+		return
+	}
+
+	err = sepsStore.Delete(b.ID().Bytes())
+	if err != nil {
+		return err
+	}
+
+	return
+}
+
+func (m *Manager) SnapshotSolidEntryPoints(lastConfirmedEpoch, latestCommitableEpoch epoch.Index, prodChan chan *SolidEntryPoints) {
+	go func() {
+		for i := lastConfirmedEpoch; i <= latestCommitableEpoch; i++ {
+			sepsPrefix := append([]byte{database.PrefixSnapshot, prefixSolidEntryPoint}, i.Bytes()...)
+			seps := make([]tangleold.BlockID, 0)
+
+			m.baseStore.IterateKeys(sepsPrefix, func(key kvstore.Key) bool {
```

Comment:
We have to remove the storage in favor of a map, but anyways: this IterateKeys will return the `sepsPrefix` as part of the key in the callback! So you are going to have weird BlockIDs deserialized, as you are consuming the prefix as part of the blockID. For this kind of operation it is better to do the IterateKeys over a storage you specialized with `WithRealm`